nilearn/mass_univariate/permuted_least_squares.py

```diff
@@ -470,43 +715,155 @@ def permuted_ols(tested_vars, target_vars, confounding_vars=None,
-    for scores_as_ranks_part in scores_as_ranks_parts:
-        scores_as_ranks += scores_as_ranks_part
-    # convert ranks into p-values
-    pvals = (n_perm + 1 - scores_as_ranks) / float(1 + n_perm)
+    (
+        vfwe_scores_as_ranks_parts,
+        vfwe_h0_parts,
+        csfwe_h0_parts,
+        cmfwe_h0_parts,
+    ) = zip(*ret)
+
+    # Voxel-level FWE
+    vfwe_h0 = np.hstack((vfwe_h0_parts))
+    vfwe_scores_as_ranks = np.zeros((n_regressors, n_descriptors))
+    for vfwe_scores_as_ranks_part in vfwe_scores_as_ranks_parts:
+        vfwe_scores_as_ranks += vfwe_scores_as_ranks_part
+
+    vfwe_pvals = (n_perm + 1 - vfwe_scores_as_ranks) / float(1 + n_perm)
+
+    # Cluster-size and cluster-mass FWE
+    csfwe_h0 = np.hstack((csfwe_h0_parts))
+    cmfwe_h0 = np.hstack((cmfwe_h0_parts))
+
+    csfwe_pvals = np.zeros_like(vfwe_pvals)
+    cmfwe_pvals = np.zeros_like(vfwe_pvals)
+
+    # TODO: Eliminate need to transpose
+    scores_original_data_4d = masker.inverse_transform(
+        scores_original_data.T
+    ).get_fdata()
+    conn = ndimage.generate_binary_structure(3, 1)
+
+    for i_regressor in range(n_regressors):
+        scores_original_data_3d = scores_original_data_4d[..., i_regressor]
+
+        # Label the clusters
+        labeled_arr3d, _ = ndimage.measurements.label(
+            scores_original_data_3d > cdt,
+            conn,
+        )
+
+        if two_sided_test:
+            # Label positive and negative clusters separately
+            n_positive_clusters = np.max(labeled_arr3d)
+            temp_labeled_arr3d, _ = ndimage.measurements.label(
+                scores_original_data_3d < cdt,
+                conn,
+            )
+            temp_labeled_arr3d[temp_labeled_arr3d > cdt] += n_positive_clusters
+            labeled_arr3d = labeled_arr3d + temp_labeled_arr3d
+            del temp_labeled_arr3d
+
+        cluster_labels, idx, cluster_sizes = np.unique(
+            labeled_arr3d,
+            return_inverse=True,
+            return_counts=True,
+        )
+        assert cluster_labels[0] == 0  # the background
+
+        # Cluster mass-based inference
+        cluster_masses = np.zeros(cluster_labels.shape)
+        for j_val in cluster_labels[1:]:  # skip background
+            cluster_mass = np.sum(
+                scores_original_data_3d[labeled_arr3d == j_val] - cdt
+            )
+            cluster_masses[j_val] = cluster_mass
+
+        p_cmfwe_vals = _null_to_p(
+            cluster_masses,
+            cmfwe_h0[i_regressor, :],
+            'upper',
+        )
+        p_cmfwe_map = p_cmfwe_vals[np.reshape(idx, labeled_arr3d.shape)]
+
+        # Convert 3D to image, then to 1D
+        cmfwe_pvals[i_regressor, :] = np.squeeze(
+            masker.transform(
+                nib.Nifti1Image(
+                    p_cmfwe_map,
+                    masker.mask_img_.affine,
+                    masker.mask_img_.header,
+                )
+            )
+        )
```

Comment:
I am not sure if this is the best way to go from 3D to 1D. I can use something like `new_img_like`, but I worry that suffers from the same limitations (i.e., only works for NiftiMaskers).